### PR TITLE
AAE-22415 Deployment message cleanup

### PR DESCRIPTION
**Pull request by revatilimaye**
### Checklist

- Jira Reference (also in PR title):
- [README](https://github.com/Alfresco/alfresco-build-tools/blob/master/docs/README.md) updated after adding/changing behaviour of an action
- Proposed version increment for [release](https://github.com/Alfresco/alfresco-build-tools/blob/master/docs/README.md#release):
  - [ ] Patch (bugfix)
  - [x] Minor (new feature)
  - [ ] Major (breaking changes)
- External PR link where changes has been tested:  https://github.com/Alfresco/hxp-studio-services/pull/2466

### Description
Issue Link: https://hyland.atlassian.net/browse/AAE-22415

### CODE REVIEW

File: .github/actions/delete-deployments/action.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+name: delete-deployments
```
Comment (atchertchian) [inline .github/actions/delete-deployments/action.yml:1]: This name should be more specific as the notion of "deployment" is vague: there are other actions handling deployment and its status, naming should be consistent with them see https://github.com/Alfresco/alfresco-build-tools?tab=readme-ov-file#github-deployment-create-and-github-deployment-status-update

File: .github/actions/delete-deployments/action.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+description: Deletes the deployments created by a workflow run in the scope of the PR.
```
Comment (atchertchian) [inline .github/actions/delete-deployments/action.yml:3]: There's nothing in the logic limiting the scope to PRs: this deletes all deployments on a given branch

File: .github/actions/delete-deployments/action.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+          page++;
+          }
+
+          await Promise.all(
+            allDeployments.map(async (deployment) => {
+              await github.rest.repos.createDeploymentStatus({
+                owner: "${{ inputs.owner }}",
+                repo: "${{ inputs.repo }}",
+                deployment_id: deployment.id,
+                state: 'inactive'
```
Comment (atchertchian) [inline .github/actions/delete-deployments/action.yml:48]: Why is this needed please? Can't we delete a deployment if it's not inactive?

File: docs/README.md [hunk 4]
```diff
@@ -395,6 +396,22 @@ Configures the git username and email to associate commits with the provided ide
...
 The two vars in the previous snippet are [workflow configuration variables](https://github.blog/changelog/2023-01-10-github-actions-support-for-configuration-variables-in-workflows/) that can be created at organization level and shared across different repositories.
 
+### delete-deployments
+
+Deletes the deployments created by a workflow run in the scope of the PR.
+
+```yaml
+  permissions:
+    deployments: write # This is required for deployment statuses management
+      - uses: Alfresco/alfresco-build-tools/.github/actions/delete-deployments@ref
```
Comment (atchertchian) [inline docs/README.md:406]: Please make it easier to copy/paste this: the omitted lines before this one are making that sample hard to navigate

**Review by atchertchian — COMMENTED** [review]
I'm not against providing utilities to delete deployments, but reading the Jira issue, I wonder how that helps with the unneeded deployments that are produced, as I don't understand why keeping deployments that we did not want to create is an issue.
Also it would be nice if we could filter on the environment name, in case not all deployments should be deleted.
But again, I'm missing the point of this cleanup.

File: .github/actions/delete-deployments/action.yml (hunk 3 above)
Comment (revatilimaye) [reply .github/actions/delete-deployments/action.yml:48]: [Here](https://docs.github.com/en/rest/deployments/deployments?apiVersion=2022-11-28#delete-a-deployment:~:text=If%20the%20repository%20has%20more%20than%20one%20deployment%2C%20you%20can%20only%20delete%20inactive%20deployments.), referring this had to set the status of deployments to inactive in order to delete.

File: .github/actions/delete-deployments/action.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
+name: delete-deployments
+
+description: Deletes the deployments created by a workflow run in the scope of the PR.
+
+inputs:
+  owner:
```
Comment (atchertchian) [inline .github/actions/delete-deployments/action.yml:6]: I don't see the point of specifying the repo here, except if we plan to call that action from a workflow that  would handle multiple repositories: using GH env variables to find the owner/repo would be better

File: .github/actions/github-deployments-delete/action.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
+name: Delete Github Deployments
+
+description: Delete gitHub deployments.Used as workaround to delete the flood of messages added to PRs when we run a workflow with a matrix that have the field environment set in the job.
+  In this case a new comment is added in PR for every entry of the matrix.
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/action.yml:4]:
```suggestion
description: >
  Deletes all GitHub deployments on a given branch.
  Used as workaround to delete the flood of messages visible on some PRs where environments are leveraged but deployments are not.
```

File: .github/actions/github-deployments-delete/action.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+inputs:
+  branch-name:
+    description: The name of the branch associated with pull request.
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/action.yml:8]:
```suggestion
    description: The name of the branch for the deployments
```

File: .github/actions/github-deployments-delete/action.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    required: true
+  environment:
+    description: The environment to delete deployments from
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/action.yml:11]:
```suggestion
    description: The optional environment for the deployments
```

File: docs/README.md [hunk 9]
```diff
@@ -683,6 +684,24 @@ jobs:
...
+```yaml
+   permissions:
+    deployments: write # This is required for deployment statuses management
+
+    steps:
+      - uses: Alfresco/alfresco-build-tools/.github/actions/github-deployments-delete@ref
+        with:
+          branch-name: ${{ github.head_ref }}
+
+```
```
Comment (atchertchian) [inline docs/README.md:703]:
```suggestion
```yaml
  permissions:
    deployments: write # This is required for deployment statuses management

  steps:
    - uses: Alfresco/alfresco-build-tools/.github/actions/github-deployments-delete@ref
      with:
        branch-name: ${{ github.head_ref }}
```
```

Comment (atchertchian) [reply docs/README.md:703]: Sorry I'm not sure the suggestion feature picked up the right things in above comment: Iwanted to fix indent (permissions and steps at same level) + remove extra blank line at the end

File: .github/actions/delete-deployments/action.yml (hunk 2 above)
Comment (revatilimaye) [reply .github/actions/delete-deployments/action.yml]: Modified the description.

File: docs/README.md (hunk 4 above)
Comment (revatilimaye) [reply docs/README.md]: Done

File: .github/actions/github-deployments-delete/action.yml (hunk 6 above)
Comment (revatilimaye) [reply .github/actions/github-deployments-delete/action.yml]: Done.

File: .github/actions/github-deployments-delete/action.yml (hunk 8 above)
Comment (revatilimaye) [reply .github/actions/github-deployments-delete/action.yml]: Done.

File: docs/README.md (hunk 9 above)
Comment (revatilimaye) [reply docs/README.md]: Corrected.

File: .github/actions/github-deployments-delete/action.yml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    description: The optional environment for the deployments
+    required: false
+
+runs:
+  using: composite
+  steps:
+    - name: delete-deployments
+      uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # v7.0.1
+      with:
+        script: |
```
Comment (gionn) [inline .github/actions/github-deployments-delete/action.yml:21]: wouldn't be better to move everything to a .js and then load it?

https://github.com/actions/github-script?tab=readme-ov-file#run-a-separate-file

should make also testing locally and lint/indent file properly

File: .github/actions/github-deployments-delete/deleteDeploymentScript.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
+module.exports = async ({github, context,ref,environment})=>{
+  const perPage = 100;
+  let page = 1;
+  let allDeployments = [];
+  while(true){
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/deleteDeploymentScript.js:5]:
```suggestion
  while(true) {
```

File: .github/actions/github-deployments-delete/deleteDeploymentScript.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        page: page,
+        environment: environment
+      });
+
+    allDeployments = [...allDeployments, ...deployments.data];
+    if (!deployments.data.length) {
+      break;
+    }
+      page++;
+    }
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/deleteDeploymentScript.js:20]:
```suggestion
    page++;
  }
```

File: .github/actions/github-deployments-delete/deleteDeploymentScript.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+          deployment_id: deployment.id,
+          state: 'inactive'
+        });
+        return github.rest.repos.deleteDeployment({
+          owner: context.repo.owner,
+          repo: context.repo.repo,
+          deployment_id: deployment.id
+        });
+      })
+    );
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/deleteDeploymentScript.js:37]:
```suggestion
  await Promise.all(
    allDeployments.map(async (deployment) => {
      await github.rest.repos.createDeploymentStatus({
        owner: context.repo.owner,
        repo: context.repo.repo,
        deployment_id: deployment.id,
        state: 'inactive'
      });
      return github.rest.repos.deleteDeployment({
        owner: context.repo.owner,
        repo: context.repo.repo,
        deployment_id: deployment.id
      });
    })
  );
```

File: .github/actions/github-deployments-delete/deleteDeploymentScript.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        environment: environment
+      });
+
+    allDeployments = [...allDeployments, ...deployments.data];
+    if (!deployments.data.length) {
+      break;
+    }
+      page++;
+    }
+
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/deleteDeploymentScript.js:21]:
```suggestion
```

File: .github/actions/github-deployments-delete/deleteDeploymentScript.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
+module.exports = async ({github, context,ref,environment})=>{
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/deleteDeploymentScript.js:1]:
```suggestion
module.exports = async ({github, context,ref,environment}) => {
```

**Review by atchertchian — APPROVED** [review]

File: .github/actions/github-deployments-delete/action.yml [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+  steps:
+    - name: delete-deployments
+      uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # v7.0.1
+      env:
+        GITHUB_ACTION_PATH: ${{ github.action_path }}
+      with:
+        script: |
+          const options={
+          ref:'${{ inputs.branch-name }}',
+          environment:'${{ inputs.environment }}'
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/action.yml:26]:
```suggestion
            ref:'${{ inputs.branch-name }}',
            environment:'${{ inputs.environment }}'
```

File: .github/actions/github-deployments-delete/action.yml [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+          };
+          const githubActionPath=process.env.GITHUB_ACTION_PATH;
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/action.yml:28]:
```suggestion
          const githubActionPath = process.env.GITHUB_ACTION_PATH;
```

File: .github/actions/github-deployments-delete/action.yml [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+          const script = require(`${githubActionPath}/deleteDeploymentScript.js`);
+          await script({
+            github,
+            context,
+            options});
```
Comment (atchertchian) [inline .github/actions/github-deployments-delete/action.yml:33]:
```suggestion
          await script({github, context, options});
```

File: .github/actions/github-deployments-delete/action.yml (hunk 16 above)
Comment (revatilimaye) [reply .github/actions/github-deployments-delete/action.yml]: Done

File: .github/actions/github-deployments-delete/action.yml (hunk 17 above)
Comment (revatilimaye) [reply .github/actions/github-deployments-delete/action.yml]: Done

File: .github/actions/github-deployments-delete/action.yml (hunk 18 above)
Comment (revatilimaye) [reply .github/actions/github-deployments-delete/action.yml]: Done

**Review by atchertchian — APPROVED** [review]
LGTM :+1:
